Topic argocd multicluster integ

leliyin opened the pull request:
VZ-8270: Support multicluster integration of Argo CD in VZ

The automation is facilitated through Rancher Proxy with creation of the k8s secret in the argocd namespace.

Summary of changes:
1. VMC CRD to include argocd registration status.
2. Cluster operator component install to create a new ‘vz-argocd-reg’ user. The Rancher API token associated with this user along with Rancher Proxy endpoint are stored in the k8s secret.
3. The new user is assigned ‘cluster-owner’ role on the cluster that the application will be deployed.
4. Cluster operator to include a configurable helm setting for the API token TTL (default 240 min).
5. The Rancher API token will be refreshed periodically if the token hasn’t been updated in 3/4 of the lifespan of the token.

Testing:
1. Manually applied vmc and verified:
- argoCDRegistration in Pending before manifest are applied
- argoCDRegistration in Completed once cluster secret is created in argocd namespace and we can deploy app in Argo CD UI
- cluster secret is refreshed with smaller ttl set and app deployment can be synced in Argo CD UI
- once vmc is deleted, cluster secret is deleted, cluster is removed in rancher
2. With CLUSTER_SYNC_ENABLED set, create cluster in rancher, verified vmc is created with argoCDRegistration status set and cluster secret created, app deployed successfully. Remove the cluster in rancher, verified cluster secret and vmc are deleted successfully.
3. Manually simulate error during secret creation, verified errors are populated with argoCDRegistration in Failed status and vmc being reconciled.

kminder commented on pkg/constants/constants.go:
```diff
@@ -174,6 +174,12 @@ const VerrazzanoClusterRancherName = "verrazzano-cluster-registrar"
 // VerrazzanoClusterRancherUsername is the username in Rancher used to identify the Verrazzano cluster user
 const VerrazzanoClusterRancherUsername = "vz-cluster-reg"
 
+// ArgoCDClusterRancherUsername is the name of secret used to grant permissions to the Verrazzano argocd user
```
Are the comments wrong or are the constant names wrong?

leliyin replied:
Have updated the comments.  We can reference 'vz-argocd-reg' as 'the Verrrazzano Argo CD cluster user' in logs as well as comments.

kminder commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +181,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+
+	headers["Content-Type"] = "application/json"
+
+	logMsg := fmt.Sprintf("Sending request message: Type: %s, URL: %s, headers: %v, payload %s", http.MethodPost, reqURL, headers, payload)
```
Could this leak credentials to the logs?

leliyin replied:
We are calling a utility function in pkg/security/password to mask password in the log message.

kminder commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +181,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+	// If the user has not been located in the response, or the status was not found, generate the user with a new password
+	pass, err := vzpassword.GeneratePassword(15)
+	if err != nil {
+		return ctx.Log().ErrorfNewErr("Failed to generate a password for the Verrazzano cluster user: %v", err)
```
Is the password for the Verrazzano cluster user?

leliyin replied:
Have fixed the comment as "Verrazzano Argo CD cluster user"

davisgomes replied:
Same comment. Please update when the changes are pushed.

kminder commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +181,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+
+	response, _, err = rancherutil.SendRequest(http.MethodPost, reqURL, headers, string(payload), rc, ctx.Log())
+	if err != nil {
+		return ctx.Log().ErrorfNewErr("Failed to create the Verrazzano cluster user in Rancher: %v", err)
```
Verrazzano cluster user?

leliyin replied:
Have fixed the log message as "Verrazzano Argo CD cluster user".

kminder commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +181,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+		return nil
+	})
+	if err != nil {
+		return ctx.Log().ErrorfNewErr("Failed create or update the Secret for the Verrazzano Cluster User: %v", err)
```
Use consistent case with other messages.

leliyin replied:
Have fixed the log message as "Verrazzano Argo CD cluster user".

kminder commented on platform-operator/helm_config/charts/verrazzano-cluster-operator/crds/clusters.verrazzano.io_verrazzanomanagedclusters.yaml:
```diff
@@ -68,6 +68,23 @@ spec:
               apiUrl:
                 description: The Verrazzano API server URL for this managed cluster.
                 type: string
+              argocdRegistration:
```
Should this match how the case ended up being handled in the VZ CR?

leliyin replied:
Good point.  I'll make the change to reflect the vmc CRD:

  status:
    argoCDRegistration:
      lastSetTimestamp: "2023-02-02T08:01:55Z"
      message: Waiting for managed cluster with id c-m-n4qrp6kw to become active before
        registering in Argo CD
      status: PendingRancherClusterRegistration

kminder commented on platform-operator/helm_config/charts/verrazzano-cluster-operator/values.yaml:
```diff
@@ -11,3 +11,6 @@ global:
   enabled: false
+
+# TTL in minutes
+argocdClusterTokenTTL: 240
```
Review for case consistency.

leliyin replied:
Done. Have updated the name as argoCDClusterTokenTTL for consistency.

kminder commented on pkg/constants/constants.go:
```diff
@@ -174,6 +174,12 @@ const VerrazzanoClusterRancherName = "verrazzano-cluster-registrar"
+// ArgoCDClusterRancherUsername is the name of secret used to grant permissions to the Verrazzano argocd user
+const ArgoCDClusterRancherName = "verrazzano-argocd-registrar"
+
+// ArgoCDClusterRancherName is the username in Rancher used to identify the Verrazzano argocd user for cluster registration
```
Review case of argocd

leliyin replied:
Done. Have updated user name as 'vz-argoCD-reg' and  sercret name as 'verrrazzano-argoCD-registrar'.

davisgomes commented on cluster-operator/apis/clusters/v1alpha1/verrazzanomanagedcluster_types.go:
```diff
@@ -129,6 +128,8 @@ type VerrazzanoManagedClusterStatus struct {
 	// The state of Rancher registration for this managed cluster.
 	RancherRegistration RancherRegistration `json:"rancherRegistration,omitempty"`
+	// The state of ArgoCD registration for this managed cluster.
+	ArgoCDRegistration ArgoCDRegistration `json:"argocdRegistration,omitempty"`
```
Should this json be argoCDRegistration?

leliyin replied:
Correct. This is already addressed.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+)
+
+func (r *VerrazzanoManagedClusterReconciler) isArgoCDEnabled() bool {
+	vz, _ := r.getVerrazzanoResource()
```
Can you handle the error for this function call?

leliyin replied:
Done.

fredtibbitts replied:
@leliyin you marked a bunch of these conversations resolved but I don't see any updates? The PR shows no commits today, did you not push the commits to your remote branch? Or maybe I'm just somehow looking at stale information?

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+}
+
+func (r *VerrazzanoManagedClusterReconciler) isRancherEnabled() bool {
+	vz, _ := r.getVerrazzanoResource()
```
Handle the error here as well.

leliyin replied:
Done

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	}
+	if vz.Status.VerrazzanoInstance == nil {
+		msg := "No instance information found in Verrazzano resource status"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr("Unable to find instance information in Verrazzano resource status")
```
Error messages should start with "Failed ..." (i.e. Failed to find instance information in Verrazzano resource status)

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	rc, err := rancherutil.NewAdminRancherConfig(r.Client, r.log)
+	if err != nil || rc == nil {
+		msg := "Could not create rancher config that authenticates with the admin user"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr(msg, err)
```
The message string does not have a "%v" for this error, so the stack trace will not be included in the error message. You could change it to this:
```
return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr(msg + ": %v", err)
```

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	}
+	secret, err := GetArgoCDClusterUserSecret(r.Client)
+	if err != nil {
+		return nil
```
`return err`?

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	secret := &corev1.Secret{}
+	nsName := types.NamespacedName{
+		Namespace: constants.VerrazzanoMultiClusterNamespace,
+		Name:      vzconst.ArgoCDClusterRancherName}
```
Can you put this end curly brace on the next line?

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+}
+
+// GetArgoCDClusterUserSecret fetches the Argo CD Verrazzano user secret
+func GetArgoCDClusterUserSecret(rdr client.Reader) (string, error) {
```
If you make this a member function (`func (r *VerrazzanoManagedClusterReconciler)`) you won't have to pass in the client reader.

leliyin replied:
Done

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	if err := rdr.Get(context.TODO(), nsName, secret); err != nil {
+		return "", err
+	}
+	return string(secret.Data["password"]), nil
```
Check that this field exists before you use it. Otherwise, you might get a nil dereference error.
```
if pw, ok := secret.Data["password"]; ok {
    return string(pw), nil
```

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	return string(secret.Data["password"]), nil
+}
+
+type TLSClientConfig struct {
+	CaData   string `json:"caData"`
+	Insecure bool   `json:"insecure"`
+}
+
+type RancherConfig struct {
+	BearerToken     string `json:"bearerToken"`
+	TLSClientConfig `json:"tlsClientConfig"`
+}
```
Can you rename the struct "`RancherConfig`" to something more unique. It will probably get confusing with two RancherConfigs with different purposes. Maybe "`ArgoCDRancherData`"?

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	TLSClientConfig `json:"tlsClientConfig"`
+}
+
+func (r *VerrazzanoManagedClusterReconciler) createOrUpdateSecret(rc *rancherutil.RancherConfig, vmc *clusterapi.VerrazzanoManagedCluster, rancherURL, clusterID string, caData []byte) (controllerutil.OperationResult, error) {
```
You don't need to return the `controllerutil.OperationResult` if you aren't using it in the function call. You can just ignore it and return the error.

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	if secret.Annotations == nil {
+		secret.Annotations = map[string]string{}
+	}
+	tokenCreated, okCreated := secret.Annotations["verrazzano.io/createTimestamp"]
+	tokenExpiresAt, okExpires := secret.Annotations["verrazzano.io/expiresAtTimestamp"]
```
Can you make these annotation strings constants since you use them twice?

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+
+	if okCreated && okExpires {
+		timeToCheck := time.Now()
+		timeCreated, err := time.Parse(time.RFC3339, tokenCreated)
+		if err != nil {
+			return err
+		}
+		timeExpired, err := time.Parse(time.RFC3339, tokenExpiresAt)
+		if err != nil {
+			return err
+		}
```
Log these errors since we are calling an imported package.

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+		if err != nil {
+			return err
+		}
+		createNewToken = (timeToCheck.Unix()-timeCreated.Unix())/(timeExpired.Unix()-timeCreated.Unix())*100 > 75
```
Can you add a comment about what we are doing here? It's a little confusing to try to read through

leliyin replied:
Done

fredtibbitts commented:
Since we do not yet have any end-to-end tests for this feature, can you please update the PR description to include how you tested this locally to make sure all of the changes work as expected? Include information on how you tested token refresh, deploying apps to managed clusters, etc.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	vz, err := r.getVerrazzanoResource()
+	if err != nil {
+		msg := "Could not find Verrazzano resource"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr("Unable to find Verrazzano resource on admin cluster: %v", err)
```
Typically we want to log errors at the source. If we are calling a function that has access to the logger, we defer to the closest we can get to the error. Please log the error in the `r.getVerrazzanoResource` function (or in any lower functions that are called from `r.getVerrazzanoResource`.

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+		if err != nil {
+			return err
+		}
+		attrs, err := rancherutil.GetTokenByName(rc, r.log, tokenName)
```
What happens if the token is not updated by the time we try to get it?

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+		BearerToken: token,
+		TLSClientConfig: TLSClientConfig{
+			CaData:   base64.StdEncoding.EncodeToString(caData),
+			Insecure: false},
```
Newline for this closed curly brace.

leliyin replied:
Done

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+		return nil
+	})
+	if err != nil {
+		return r.log.ErrorfThrottledNewErr("Failed configuring %s %s: %s", ClusterRoleTemplateBindingKind, nsn.Name, err.Error())
```
Use a `%v` and pass in the error so that we can get a stack trace.

leliyin replied:
Done

davisgomes commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+			Namespace: constants.ArgoCDNamespace,
+		},
+	}
+	if err := r.Delete(context.TODO(), &clusterSec); err != nil {
```
You can use `client.IgnoreNotFound(err)` i.e.
```
if err := r.Delete(context.TODO(), &clusterSec); client.IgnoreNotFound(err) != nil {
    return r.log.ErrorfNewErr("Failed to delete Argo CD cluster secret", err)
}
```

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,359 @@
+//	WHEN the secret annotation createTimestamp/expiresAtTimestamp is x(s) and x+4(s) respectively
+//	and we sleep for 4(s)
+//	THEN we obtain new token and the annotation createTimestamp/expiresAtTimestamp are updated accordingly
+func TestMutateClusterSecretWithRefresh(t *testing.T) {
```
There's a lot of copy paste here. I think it's okay for a unit test, but you could consider using the loop testing format: https://dave.cheney.net/2013/06/09/writing-table-driven-tests-in-go

desagar replied:
Similar to the TestUpdateArgoCDClusterRoleBindingTemplate below, which has several cases.

davisgomes commented on cluster-operator/controllers/vmc/vmc_controller.go:
```diff
@@ -226,8 +226,36 @@ func (r *VerrazzanoManagedClusterReconciler) doReconcile(ctx context.Context, lo
 
+	log.Debugf("Registering ArgoCD for VMC %s", vmc.Name)
+	var argoCDRegistration *clustersv1alpha1.ArgoCDRegistration
+	if r.isArgoCDEnabled() {
+		if r.isRancherEnabled() {
+			argoCDRegistration, err = r.registerManagedClusterWithArgoCD(vmc)
+			if err != nil {
+				r.handleError(ctx, vmc, "Failed to register managed cluster with Argo CD", err, log)
+				_ = r.updateStatus(ctx, vmc, func(vmc *clustersv1alpha1.VerrazzanoManagedCluster) {
+					if argoCDRegistration != nil {
+						vmc.Status.ArgoCDRegistration = *argoCDRegistration
+					}
+				})
+				return newRequeueWithDelay(), err
+			}
+		} else {
+			now := metav1.Now()
+			vmc.Status.ArgoCDRegistration = clustersv1alpha1.ArgoCDRegistration{
+				Status:    clustersv1alpha1.RegistrationPendingRancher,
+				Timestamp: &now,
+				Message:   "Waiting for Rancher managed cluster to become active"}
+		}
+	}
+
```
These nested `if`s are a bit ugly. Here's how I think you could format it:
```
argoCDEnabled := r.isArgoCDEnabled()
rancherEnabled := r.isRancherEnabled()
if argoCDEnabled && rancherEnabled {
    // register ArgoCD
}
if !rancherEnabled && argoCDEnabled {
  // update status
}
```

leliyin replied:
Done.

davisgomes commented on cluster-operator/controllers/vmc/vmc_controller.go:
```diff
@@ -226,8 +226,36 @@ func (r *VerrazzanoManagedClusterReconciler) doReconcile(ctx context.Context, lo
+		if r.isRancherEnabled() {
+			argoCDRegistration, err = r.registerManagedClusterWithArgoCD(vmc)
+			if err != nil {
+				r.handleError(ctx, vmc, "Failed to register managed cluster with Argo CD", err, log)
+				_ = r.updateStatus(ctx, vmc, func(vmc *clustersv1alpha1.VerrazzanoManagedCluster) {
+					if argoCDRegistration != nil {
+						vmc.Status.ArgoCDRegistration = *argoCDRegistration
+					}
+				})
```
I don't think you need to update the status like you have here. You can follow suit from the push manifest block and create a new function `r.setStatusConditionArgoCDRegistered`

davisgomes commented on cluster-operator/controllers/vmc/vmc_controller.go:
```diff
@@ -226,8 +226,36 @@ func (r *VerrazzanoManagedClusterReconciler) doReconcile(ctx context.Context, lo
+
 	r.setStatusConditionReady(vmc, "Ready")
-	statusErr := r.updateStatus(ctx, vmc)
+	statusErr := r.updateStatus(ctx, vmc, func(vmc *clustersv1alpha1.VerrazzanoManagedCluster) {
+		if argoCDRegistration != nil {
+			vmc.Status.ArgoCDRegistration = *argoCDRegistration
+		}
+	})
```
Same comment about the status update. I don't think you need to add this parameter.

davisgomes commented on pkg/rancherutil/rancher_config.go:
```diff
@@ -185,6 +189,76 @@ func getUserToken(rc *RancherConfig, log vzlog.VerrazzanoLogger, secret, usernam
+	TTL       int    `json:"ttl"`
+}
+
+// SetTokenTTL creates a user token with ttl
```
What granularity is the ttl at. Hours?

leliyin replied:
Have updated the comment that ttl is in minutes.

davisgomes commented on pkg/rancherutil/rancher_config.go:
```diff
@@ -185,6 +189,76 @@ func getUserToken(rc *RancherConfig, log vzlog.VerrazzanoLogger, secret, usernam
+		return nil, err
+	}
+
+	err = httputil.ValidateResponseCode(response, http.StatusOK)
+	if err != nil {
+		return nil, err
+	}
+
+	jsonString, err := gabs.ParseJSON([]byte(responseBody))
+	if err != nil {
+		return nil, err
+	}
```
Log these errors

leliyin replied:
Done.

davisgomes commented on pkg/rancherutil/rancher_config.go:
```diff
@@ -185,6 +189,76 @@ func getUserToken(rc *RancherConfig, log vzlog.VerrazzanoLogger, secret, usernam
+	if err != nil {
+		return "", "", err
+	}
+	jsonString, err := gabs.ParseJSON([]byte(responseBody))
```
I think you can json unmarshal this into a struct like the ones you defined above

leliyin replied:
Done.

davisgomes commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -83,7 +84,14 @@ func (c clusterOperatorComponent) postInstallUpgrade(ctx spi.ComponentContext) e
+		if err := rancher.CreateOrUpdateRoleTemplate(ctx, vzconst.VerrazzanoClusterRancherName); err != nil {
+			return err
+		}
+		if vzcr.IsArgoCDEnabled(ctx.EffectiveCR()) {
+			if err := createVZArgoCDUser(ctx); err != nil {
+				return err
+			}
+		}
```
You can reverse this to remove the nested ifs:
```
if !vzcr.IsArgoCDEnabled(ctx.EffectiveCR()) {
    return nil
}
return := createVZArgoCDUser(ctx)

leliyin replied:
Done.

davisgomes commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +181,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+
+	logMsg := fmt.Sprintf("Sending request message: Type: %s, URL: %s, headers: %v, payload %s", http.MethodPost, reqURL, headers, payload)
+	// Mask the bearer token in the log message
+	logMsg = vzpassword.MaskFunction("Authorization:Bearer ")(logMsg)
```
Should you have a space between the colon and Bearer?

leliyin replied:
I checked the same was done for createVZClusterUser.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+	err = r.updateArgoCDClusterRoleBindingTemplate(vmc)
+	if err != nil {
+		msg := "Failed to update Argo CD ClusterRoleBindingTemplate"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr(msg, err)
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,303 @@
+
+	// Create or update on the local cluster
+	return controllerruntime.CreateOrUpdate(context.TODO(), r.Client, &secret, func() error {
+		err := r.mutateClusterSecret(&secret, rc, vmc.Name, clusterID, rancherURL, caData)
+		if err != nil {
+			return err
+		}
+		return nil
```
done.

leliyin commented on pkg/rancherutil/rancher_config.go:
```diff
@@ -185,6 +189,76 @@ func getUserToken(rc *RancherConfig, log vzlog.VerrazzanoLogger, secret, usernam
+		ClusterID: clusterID,
+		TTL:       val * 60000,
+	}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return "", "", err
+	}
+	action := http.MethodPost
+	reqURL := rc.BaseURL + tokensPath
+	headers := map[string]string{"Authorization": "Bearer " + rc.APIAccessToken, "Content-Type": "application/json"}
+
+	response, responseBody, err := SendRequest(action, reqURL, headers, string(data), rc, log)
+	if err != nil {
+		return "", "", err
+	}
+	err = httputil.ValidateResponseCode(response, http.StatusCreated)
+	if err != nil {
+		return "", "", err
+	}
+	jsonString, err := gabs.ParseJSON([]byte(responseBody))
+	if err != nil {
+		return "", "", err
+	}
```
Done.

leliyin commented:
> Since we do not yet have any end-to-end tests for this feature, can you please update the PR description to include how you tested this locally to make sure all of the changes work as expected? Include information on how you tested token refresh, deploying apps to managed clusters, etc.

I've updated the PR with tests I performed locally. We are adding e2e/multiclusters tests as we speak, the PR will be shortly followed.

I have just merged first round of review comments from Kevin and Davis. Please review.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+	if createNewToken {
+		// Obtain a new token with ttl set using bearer token obtained
+		ttl := os.Getenv(argocdClusterTokenTTLEnvVarName)
+		newToken, tokenName, err := rancherutil.SetTokenTTL(rc, r.log, ttl, clusterID)
```
Correct. The Rancher API to set the token TTL returns create timestamp, not expires timestamp, hence we need to follow up with GET with the token name to get the expires timestamp.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr("Unable to find instance information in Verrazzano resource status")
+	}
+
+	var rancherURL = *(vz.Status.VerrazzanoInstance.RancherURL) + k8sClustersPath + clusterID
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+	return newArgoCDRegistration(clusterapi.MCRegistrationCompleted, msg), nil
+}
+
+// argocdClusterAdd registers cluster using the Rancher Proxy by creating a user in rancher, with api token and cluster roles set, and a secret containing Rancher proxy for the cluster
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+}
+
+// GetArgoCDClusterUserSecret fetches the Argo CD Verrazzano user secret
+func (r *VerrazzanoManagedClusterReconciler) GetArgoCDClusterUserSecret() (string, error) {
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+	if pw, ok := secret.Data["password"]; ok {
+		return string(pw), nil
+	}
+	return "", nil
```
Good catch. Fixed.

leliyin commented on pkg/constants/constants.go:
```diff
@@ -174,6 +174,12 @@ const VerrazzanoClusterRancherName = "verrazzano-cluster-registrar"
 const VerrazzanoClusterRancherUsername = "vz-cluster-reg"
 
+// ArgoCDClusterRancherName is the name of secret for the Verrazzano Argo CD cluster user
+const ArgoCDClusterRancherName = "verrazzano-argoCD-registrar"
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,359 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "demo" + "-" + clusterSecretName,
+			Namespace:   constants.ArgoCDNamespace,
+			Annotations: map[string]string{"verrazzano.io/createTimestamp": time.Now().Format(time.RFC3339), "verrazzano.io/expiresAtTimestamp": time.Now().Add(4 * time.Second).Format(time.RFC3339)},
```
Good idea. Have set expiration to +10hr.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+			return r.log.ErrorfNewErr("Failed to parse expired timestamp: %v", err)
+		}
+		// Obtain new token if the time elapsed between time created and expired is greater than 3/4 of the lifespan of the token
+		createNewToken = (timeToCheck.Unix()-timeCreated.Unix())/(timeExpired.Unix()-timeCreated.Unix())*100 > 75
```
Done.

leliyin commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,359 @@
+	assert.NoError(t, err)
+
+	err = r.mutateClusterSecret(secret, rc, vmc.Name, clusterID, rancherURL, caData)
+	assert.NoError(t, err)
```
Good point. Have added an assert to check the token from the secret is same as the token set in the unit-test.

leliyin commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,359 @@
+	mocker := gomock.NewController(t)
+	httpMock := mocks.NewMockRequestSender(mocker)
+	// Expect an HTTP request to fetch the token from Rancher only
+	httpMock.EXPECT().
```
Correct. this mock expectation is from call to NewRancherConfigForUser to get the Rancher Config to pass to mutateClusterSecret. As this test case is to assert token not refreshed, I've added an assert to check the token is not changed after mutateClusterSecret completes. So we still need call to NewRancherConfigForUser, yes?

fredtibbitts replied:
I think having `NewRancherConfigForUser` and the associated code isn't needed for this unit test, it just adds unnecessary complexity. I don't feel strongly about removing it so if you want to leave it, that's fine.

leliyin commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,359 @@
+	rc, err := rancherutil.NewRancherConfigForUser(cli, constants.ArgoCDClusterRancherUsername, "foobar", log)
+	assert.NoError(t, err)
+
+	time.Sleep(4 * time.Second)
```
Good idea. Updated code per suggestion and removed the sleep.

leliyin commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,309 @@
+const (
+	clusterSecretName               = "cluster-secret"
+	argocdClusterTokenTTLEnvVarName = "ARGOCD_CLUSTER_TOKEN_TTL" //nolint:gosec
+	createTimestamp                 = "verrazzano.io/createTimestamp"
```
I've incorporated Fred's suggestion for annotation naming.  We can update further if needed.

fredtibbitts commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -126,19 +130,20 @@ func (r *VerrazzanoManagedClusterReconciler) createClusterSecret(vmc *clusterapi
 		return string(pw), nil
 	}
-	return "", nil
+	return "", err
```
`err` is `nil` here (you check for != `nil` above and return so it must be `nil`). You need to create an error here and return it.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+
+	// If the managed cluster is not active, we should not attempt to register in Argo CD
+	rc, err := rancherutil.NewAdminRancherConfig(r.Client, r.log)
+	if err != nil || rc == nil {
```
this rc == nil check is not needed since if err is nil we will get back an rc. The logging in this case can also be skipped.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	err = r.updateArgoCDClusterRoleBindingTemplate(vmc)
+	if err != nil {
+		msg := "Failed to update Argo CD ClusterRoleBindingTemplate"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr(msg+": %v", err)
```
any logs with ErrorfNewErr should be formatted as follows rather than using string concatenation.

r.log.ErrorfNewErr("%s: %v", msg, err)

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+
+	err = r.updateArgoCDClusterRoleBindingTemplate(vmc)
+	if err != nil {
+		msg := "Failed to update Argo CD ClusterRoleBindingTemplate"
```
The called function logs the err so this one does not need a log msg

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+		msg := "Failed to create Argo CD cluster secret"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr("Unable to create cluster secret on admin cluster: %v", err)
+	}
+	msg := "Successfully registered managed cluster in ArgoCD"
```
General comment for this function and all functions - any logs with ErrorfNewErr (or anything with an "f" suffix for the logging) should be formatted as follows rather than using string concatenation.

r.log.ErrorfNewErr("%s: %v", msg, err)

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	err = r.createClusterSecret(vmc, clusterID, rancherURL)
+	if err != nil {
+		msg := "Failed to create Argo CD cluster secret"
+		return newArgoCDRegistration(clusterapi.MCRegistrationFailed, msg), r.log.ErrorfNewErr("Unable to create cluster secret on admin cluster: %v", err)
```
The called function logs the err so this one does not need a log msg

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+
+	caCert, err := common.GetRootCA(r.Client)
+	if err != nil {
+		return err
```
this err case should log here since the GetRootCA function doesn't take a logger.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	TLSClientConfig `json:"tlsClientConfig"`
+}
+
+func (r *VerrazzanoManagedClusterReconciler) createOrUpdateSecret(rc *rancherutil.RancherConfig, vmc *clusterapi.VerrazzanoManagedCluster, rancherURL, clusterID string, caData []byte) error {
```
This function should be called createOrUpdateArgoCDSecret, to be more clear. It is declared on the reconciler object and that object is used for both argo and non-argo work.
Also a function comment at the top would be helpful.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	return err
+}
+
+func (r *VerrazzanoManagedClusterReconciler) mutateClusterSecret(secret *corev1.Secret, rc *rancherutil.RancherConfig, clusterID, cluserName string, rancherURL string, caData []byte) error {
```
This function should be called mutateArgoCDSecret, to be more clear. It is declared on the reconciler object and that object is used for both argo and non-argo work.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+}
+
+// createClusterSecret registers cluster using the Rancher Proxy by creating a user in rancher, with api token and cluster roles set, and a secret containing Rancher proxy for the cluster
+func (r *VerrazzanoManagedClusterReconciler) createClusterSecret(vmc *clusterapi.VerrazzanoManagedCluster, clusterID, rancherURL string) error {
```
This function should be called createArgoCDSecret, to be more clear. It is declared on the reconciler object and that object is used for both argo and non-argo work.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	return newArgoCDRegistration(clusterapi.MCRegistrationCompleted, msg), nil
+}
+
+// createClusterSecret registers cluster using the Rancher Proxy by creating a user in rancher, with api token and cluster roles set, and a secret containing Rancher proxy for the cluster
```
this comment should be clarified e.g. the rancher user for argoCD must already exist at this point - if that's the case, it's not created by this function.

createArgoCDClusterSecret registers cluster with ArgoCD using the "vz-argoCD-reg" user and the Rancher proxy URL for the cluster.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+)
+
+const (
+	clusterSecretName               = "cluster-secret"
```
This constant should be named with a prefix indicating it's related to argo. 
Likewise, it would be helpful for the value to have "argoCD" in it (e.g. "argoCD-cluster-secret").

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	}
+	data, err := json.Marshal(rancherConfig)
+	if err != nil {
+		return err
```
We should log this error before returning it.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+	return nil
+}
+
+// updateRancherClusterRoleBindingTemplate creates a new ClusterRoleBindingTemplate for the given VMC
```
update comment to correct the function name.

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+		Version: APIGroupVersionRancherManagement,
+		Kind:    UserListKind,
+	})
+	err := r.Client.List(context.TODO(), &usersList, &client.ListOptions{})
```
you can just do r.List() instead of r.Client.List here

desagar commented on cluster-operator/controllers/vmc/argocd.go:
```diff
@@ -0,0 +1,314 @@
+		},
+	}
+	if err := r.Delete(context.TODO(), &clusterSec); client.IgnoreNotFound(err) != nil {
+		return r.log.ErrorfNewErr("Failed to delete Argo CD cluster secret", err)
```
missing %v in the format string to print the error in the log msg

desagar commented on cluster-operator/controllers/vmc/argocd_test.go:
```diff
@@ -0,0 +1,365 @@
+		})
+
+	// Expect an HTTP request to fetch the token from Rancher
+	httpMock.EXPECT().
```
this mock expectation can be moved to its own function e.g. expectRancherLoginRequest(), and then shared by both of the test functions.

desagar commented on cluster-operator/controllers/vmc/vmc_controller.go:
```diff
@@ -226,8 +226,35 @@ func (r *VerrazzanoManagedClusterReconciler) doReconcile(ctx context.Context, lo
+	}
+	if argoCDEnabled && rancherEnabled {
+		argoCDRegistration, err = r.registerManagedClusterWithArgoCD(vmc)
+		vmc.Status.ArgoCDRegistration = *argoCDRegistration
```
This line should be _after_ the error check. We should not dereference *argoCDRegistration if there was an error.

desagar commented on cluster-operator/controllers/vmc/vmc_controller.go:
```diff
@@ -226,8 +226,35 @@ func (r *VerrazzanoManagedClusterReconciler) doReconcile(ctx context.Context, lo
+		vmc.Status.ArgoCDRegistration = clustersv1alpha1.ArgoCDRegistration{
+			Status:    clustersv1alpha1.RegistrationPendingRancher,
+			Timestamp: &now,
+			Message:   "Waiting for Rancher managed cluster to become active"}
```
This message should say something about Rancher not being enabled, since that is the reason we can't do the argo registration.

desagar commented on pkg/rancherutil/rancher_config.go:
```diff
@@ -185,6 +188,79 @@ func getUserToken(rc *RancherConfig, log vzlog.VerrazzanoLogger, secret, usernam
+func GetTokenByName(rc *RancherConfig, log vzlog.VerrazzanoLogger, tokenName string) (*TokenAttrs, error) {
+	action := http.MethodGet
+	reqURL := rc.BaseURL + tokensPath + "/" + tokenName
+	headers := map[string]string{"Authorization": "Bearer " + rc.APIAccessToken, "Content-Type": "application/json"}
```
why is the Content-Type needed for this GET request that doesn't send a body?

desagar commented on platform-operator/controllers/verrazzano/component/clusteroperator/cluster_operator.go:
```diff
@@ -173,3 +180,88 @@ func constructVZUserJSON(pass string) ([]byte, error) {
+	return nil
+}
+
+func constructVZArgoUserJSON(pass string) ([]byte, error) {
```
not as big of a deal but this can be combined easily as well with the other construct JSON function

desagar commented on cluster-operator/apis/clusters/v1alpha1/verrazzanomanagedcluster_types.go:
```diff
@@ -95,23 +95,22 @@ type RancherRegistration struct {
-	Timestamp string `json:"lastSetTimestamp"`
+	// +optional
+	Timestamp *metav1.Time `json:"lastSetTimestamp,omitempty"`
 	// Supporting message related to the Rancher registration status
```
This comment needs to say ArgoCD registration instead of Rancher registration. This also ends up in the CRD so you'll need to rerun "make manifests"

leliyin commented:
@desagar @davis-gomes @fredtibbitts @kminder  I've addressed all the review comments. Please let me know if the review is still pending or we can have a merge to the master branch
